node/src/p2p/swarm_manager.rs

```diff
@@ -0,0 +1,703 @@
+            // Change keep alive of ongoing connections to `true`
+            for conn_id in self.peer_tracker.connections(peer_id) {
+                self.swarm
+                    .behaviour_mut()
+                    .connection_control
+                    .set_keep_alive(peer_id, conn_id, true);
+            }
+        }
+    }
+
+    fn unprotect(&mut self, peer_id: &PeerId, tag: u32) {
+        if self.peer_tracker.unprotect(peer_id, tag) {
+            // Change keep alive of ongoing connections to `false`
+            for conn_id in self.peer_tracker.connections(peer_id) {
+                self.swarm
+                    .behaviour_mut()
+                    .connection_control
+                    .set_keep_alive(peer_id, conn_id, false);
+            }
+        }
+    }
+
+    async fn peer_health_check(&mut self) {
+        let info = self.peer_tracker.info();
+        let protected_full_nodes = self.peer_tracker.protected_len(FULL_PROTECT_TAG);
+        let protected_archival_nodes = self.peer_tracker.protected_len(ARCHIVAL_PROTECT_TAG);
+
+        if info.num_connected_peers == 0 {
+            warn!("All peers disconnected");
+            self.bootstrap();
+        }
+
+        // Based on the spec, when a protected node gets disconnected, the we unprotect
+        // it (check `on_peer_disconnected`), and another node must be choosen to be
+        // protected.
+        if protected_full_nodes < MAX_PROTECTED_FULL_NODES {
+            // Protect the best full nodes to fill up the gap.
+            self.protect_best_peers(
+                MAX_PROTECTED_FULL_NODES - protected_full_nodes,
+                FULL_PROTECT_TAG,
+                Peer::is_full,
+            );
+
+            // If we still have less than max, we initiate full node discovery.
+            if self.peer_tracker.protected_len(FULL_PROTECT_TAG) < MAX_PROTECTED_FULL_NODES {
+                self.start_full_node_kad_query();
+            }
+        }
+
+        if protected_archival_nodes < MAX_PROTECTED_ARCHIVAL_NODES {
+            // Protect the best archival nodes to fill up the gap.
+            self.protect_best_peers(
+                MAX_PROTECTED_ARCHIVAL_NODES - protected_archival_nodes,
+                ARCHIVAL_PROTECT_TAG,
+                Peer::is_archival,
+            );
+
+            // If we still have less than max, we initiate archival node discovery.
+            if self.peer_tracker.protected_len(ARCHIVAL_PROTECT_TAG) < MAX_PROTECTED_ARCHIVAL_NODES
+            {
+                self.start_archival_node_kad_query();
+            }
+        }
+    }
+
+    /// Protect the best N peers based on their ping latency.
+    fn protect_best_peers<F>(&mut self, count: usize, tag: u32, condition: F)
+    where
+        F: Fn(&Peer) -> bool,
+    {
+        // Select peers that:
+        //
+        // * Satisfy `condition`
+        // * Are connected
+        // * Not already protected for `tag`
+        // * Have measured ping latency
```

Comment:
```suggestion
        // * Have known ping latency
```
above makes it sound as if the peers were doing the measuring ;p